Fixtures homework.

LR100days opened the pull request:
Hi, Artem!

Once again, thank you for the help with API request!

I have a question about using the faker library. In my tests I use 2 approaches:

1. Use `HelperBase` class that extends page objects classes, if page object file already exists;
2. Just importing faker library and use it directly in the test (like in this `testWithFixture.spec.ts` test or in `performingApiRequest.spec.ts` test).

The question is if it's ok to use it this way, like in the second approach, or it's better to import the `HelperBase class` itself, and then create an object of this class and use it in tests, instead of direct call to faker library? 
Could you please advise what would be better for the cases mentioned in the second approach?

ba-instructor reviewed (approved):
Looks good, you have figured out the fixtures.

What I would also suggest is to refactor the API calls. They take a lot of space and hard to read. So it would be much better to create some kind of "ApiHelper" class and create different API method that would do some operations for you over API.
Like: createOner(), addPetToOwner(), addVisitsForPet(), deleteOwnerById() and so on. Then you can call those methods in the fixture and the script will be much cleaner and easier to read. Also those helper methods can be used anytime in other test, whenever you need to perform any of those operations, you have API method ready :)

LR100days commented:
Artem, thank you so much for those explanations!
I refactored my code to use new class ApiHelper and also updated some tests to use HelperBase instead of faker library.

Could you please validate those changes? 
Thanks in advance :)

ba-instructor reviewed (commented):
Please check the comments

LR100days commented on tests/workingWithAPI.spec.ts:
```diff
@@ -90,7 +90,7 @@ test('add and delete an owner', async ({ page, request }) => {
   await pm.onOwnersPage().validateOwnerDetailsInTable(`${randomOwnerFirstName} ${randomOwnerLastName}`, randomOwnerAddress, randomOwnerCity, randomOwnerTelephone)
   
-  const apiHelper = new ApiHelper();
+  const apiHelper = new ApiHelper(page);
```
Thanks for this comment. You're right, I missunderstood this. Updated the code.

LR100days commented on test-options.ts:
```diff
@@ -16,16 +16,17 @@ export type TestOptions = {
         
-        const apiHelper = new ApiHelper();
+        const apiHelper = new ApiHelper(page);
         const ownerID = await apiHelper.createOwnerByApi(request, randomOwnerFirstName, randomOwnerLastName)
```
Totally agree! This makes much more sense now, thanks. Updated the code.

LR100days commented:
Thanks for the comments and suggestions how to impove my code.  I tried to fix that, could you please check if my code fix is sufficient? 
Thanks!

ba-instructor commented on page_objects/apiHelper.ts:
```diff
@@ -1,26 +1,33 @@
-            super(page)
-        }
+    readonly request: APIRequestContext;
+    readonly page: Page
```
I don't like that you have a property that does not have value assigned in the constructor. So you pass the empty `this.page` to a Helper base just to make the compiler happy. Not very clean...

Currently helper base has only methods related to data generation. And in the tests, you call it directly through the page objects, since it's a base class for page objects. At the same time, HelperBase called directly in the fixtures. This creates a mix how this class is used without clear boundaries of responsibilities.

A helperBase class is usually needed for the methods shared across different page object. Currently it's not a case for you. From what I see, none of the page objects use methods from HelperBase. You use those methods directly in the tests.

In this case, it would make more sense, instead of "HelperBase" to create a new utility class like "dataGenerationHelper", that will be responsible only for different data generation. It will not have a dependency on page fixture or any other fixture. You can import this helper directly in the test or page object when needed.

It will make the structure more clear, where the data is coming from, and you will not need to "hack" the passing the empty page property to a HelperBase.

You can keep the HelperBase in your project, it will be just empty with just a constructor and you can use it later if you have some methods shared across page objects.
And DataGenerationHelper will be responsible for just data generation where needed.

ba-instructor replied:
And since now you will have two helpers, I think it's also a good idea to move apiHelper and dataGenerationHelper to a separate folder like, "utils" from the "page_objects" folder, again to separate the responsibility and better organize files in the project. If in future you will have more "helpers" they will have a dedicated folder

ba-instructor reviewed (commented):
Please check the comment )

LR100days commented:
Thanks, Artem, for the simplified and logical example of my code structure! I updated the files (hopefully according to your comments), only one question remains - do I also need to move the HelperBase file to the Utils folder or is it usually located in the pageObjects folder?
